src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -72,17 +75,32 @@ internal ZipArchiveEntry(ZipArchive archive, ZipCentralDirectoryFileHeader cd)
             _everOpenedForWrite = false;
             _outstandingWriteStream = null;
 
-            FullName = DecodeEntryName(cd.Filename);
+            FullName = DecodeBytesToString(cd.Filename);
 
             _lhUnknownExtraFields = null;
-            // the cd should have these as null if we aren't in Update mode
+            // the cd should have this as null if we aren't in Update mode
             _cdUnknownExtraFields = cd.ExtraFields;
-            _fileComment = cd.FileComment;
+
+            if (cd.FileComment.Length > 0)
+            {
+                _fileComment = cd.FileComment;
+                _fileCommentAsString = ZipHelper.TruncateEncodedBytesAndReturnString(
+                    ref _fileComment,
+                    _archive.EntryNameAndCommentEncoding ?? Encoding.UTF8,
+                    ushort.MaxValue);
+            }
+            else
+            {
+                _fileCommentAsString = null;
+                _fileComment = null;
+            }
 
             _compressionLevel = null;
+
+            _hasUnicodeEntryName = _hasUnicodeComment = IsGeneralPurposeUnicodeFlagSet();
```

Comment:
these fields seem to always have the same value, why don't we have a single field?

Moreover, `IsGeneralPurposeUnicodeFlagSet` seems to be a very simple method, what do we gain by storing it's result in a field?

Reply:
This is the constructor that loads the entries from an _existing_ archive. Whatever encoding value we have in the current entry header, needs to affect both strings.

If the user creates an archive, there are two places where the encoding value can be changed: the entry name and the comment.

If the user sets an ascii entry name, but an unicode comment, then the entry encoding bit needs to be saved as unicode.

If the user sets an unicode entry name, but an ascii comment, then the bit needs to be saved as unicode.

If both are set to ascii, the bit is ascii.

The entry name can only be set once, there's no chance to change it.

So there's the case where the user updates the comment. If the entry was ascii, and the comment was originally saved as unicode, but then updated via the setter to an ascii string, then the encoding bit should be set to false because both booleans were switched to false.

Reply:
But thinking about this again, it isn't necessary to save two booleans. Once any of the two strings is detected as utf8, I can simply keep saving the encoding bit as utf8, period.

Reply:
> Moreover, IsGeneralPurposeUnicodeFlagSet seems to be a very simple method, what do we gain by storing it's result in a field?

The logic is used in two places, so yes. I want to save it as a method to avoid repeating code.